### PR TITLE
CMake/Ninja: fix dependencies for .cu files

### LINKED ISSUE #26304: When CUDA is enabled, recompiling after editing aten/src/ATen/core/dispatch/DispatchTable.h doesn't result in all necessary files getting recompiled
Other files may also be susceptible, but the issue reproduces 100% on `aten/src/ATen/core/dispatch/DispatchTable.h`

Steps to reproduce:
1. Build PyTorch `python setup.py develop`
2. Edit the "Didn't find kernel to dispatch to" error message to insert some string. Example diff:

```
diff --git a/aten/src/ATen/core/dispatch/DispatchTable.h b/aten/src/ATen/core/dispatch/DispatchTable.h
index 7b4af315f1..3a5879cbda 100644
--- a/aten/src/ATen/core/dispatch/DispatchTable.h
+++ b/aten/src/ATen/core/dispatch/DispatchTable.h
@@ -263,7 +263,7 @@ private:
       }
 
       const std::string dispatch_key_str = dispatch_key.has_value() ? toString(*dispatch_key) : "None";
-      TORCH_CHECK(false, "Didn't find kernel to dispatch to for operator '", operator_name_,
+      TORCH_CHECK(false, "ARGLEBARGLEFARR Didn't find kernel to dispatch to for operator '", operator_name_,
                "'. Tried to look up kernel for dispatch key '", dispatch_key_str,
                "'. Registered dispatch keys are: ", listAllDispatchKeys());
   }
```

3. Rebuild PyTorch `python setup.py develop`
4. Run `python -c "import torch; torch.cat([], dim=1)`

Expected result: you see ARGLEBARGLEFARR
Actual result: you don't see it

I did an experiment where I did one of these botched recompiles, and then compared the set of files which had ARGLEBARGLEFARR with the set of files that had the normal error message. What I found is that CUDA generated files had not been recompiled, but CPU (and final library) all had the string. The fact that the CUDA files are not recompiling seems problematic; however, I don't understand how this could have caused the problem, as the test I made above is CPU (perhaps some sort of weak symbol situation?)

cc @ezyang @gchanan @zou3519 @smessmer

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #26304 

After this patch `build.ninja` entries for `.cu` files will contain a `depfile` variable pointing to a `.NVCC-depend` file containing dependencies (i.e., header files included directly or indirectly) of the `.cu` source file. Until now, those `.NVCC-depend` files were being transposed into `.cu.o.depend` files in CMake format. That did not work as intended because the `.cu.o` target file was declared to be dependent on the `.cu.o.depend` file itself, rather than its contents. In fact, Ninja lacks the functionality to process dependencies in the CMake format of those `.cu.o.depend` files.

This was tested on Linux as described in https://github.com/pytorch/pytorch/issues/26304#issuecomment-614667170
I have also verified that the original problem does not reproduce with Makefiles (i.e., when `ninja` is not present in the system) and that PyTorch still build successfully with Makefiles after this patch.

### CODE REVIEW

**Comment by ezyang** [comment]
WOOOOW! This is super awesome!

**Comment by ezyang** [comment]
We should upstream this fix to cmake

**Review by ezyang — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Comment by Baranowski** [comment]
Thanks, I'm glad you like it.

> We should upstream this fix to cmake

Sure, do you want me to do that?

**Comment by malfet** [comment]
Hmm, looks like this change make CUDA files dirty all the time:
```
$ ninja torch_python
[22/333] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/operators/torch_cuda_generated_segment_reduction_op_gpu.cu.o
/home/nshulga/git/pytorch/cmake/../third_party/eigen/Eigen/src/Core/arch/GPU/PacketMathHalf.h(149): warning: missing return statement at end of non-void function "Eigen::internal::ptrue(const Packet &) [with Packet=half2]"

[30/333] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/torch_cuda_generated_THCTensorMath.cu.o^C
ninja: build stopped: interrupted by user.
$ ninja -d explain torch_python
[0/333] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathPointwiseFloat.cu.o
```

**Comment by ezyang** [comment]
I'm going to unland it for now.

**Comment by Baranowski** [comment]
Sorry, should have tested that. I will fix this.

**Comment by malfet** [comment]
@Baranowski, I found `ninja -d explain` is a useful tool at understanding why ninja think rebuild is warranted.
